File: recursividade/javascript/conta-caracteres.js
function charCount(ch, text) {
  return charCountTailRecursive(ch.toLowerCase(), text.toLowerCase(), 0, 0);
}

function charCountTailRecursive(ch, text, index, count) {
  if (index >= text.length) {
    return count;
  }

  const newCount = (text[index] == ch) ? count + 1 : count;

  return charCountTailRecursive(ch, text, index + 1, newCount);
}

const result = charCount("b", "Batata para o bebê");
console.log(result);

File: recursividade/java/src/ContaCaracteres.java
public class ContaCaracteres {

	public static void main(String[] args) {
        System.out.println(charCount('b', "Batata para o bebê"));
    }

    public static int charCount(char ch, String text) {
        return charCountTailRecursive(Character.toLowerCase(ch), text.toLowerCase(), 0, 0);
    }

    private static int charCountTailRecursive(char ch, String text, int index, int count) {
        if (index >= text.length()) {
            return count;
        }

        int newCount = (text.charAt(index) == ch) ? count + 1 : count;
        return charCountTailRecursive(ch, text, index + 1, newCount);
    }
}
